File: app/providers/formats/lrc.js
// LRC parsing → canonical timeline with line and (interpolated) word timings.

import { wordsAcrossSpan } from './estimate.js';

// Estimate a word's syllable count — a much better proxy for how long it's sung
// than raw character length (the previous heuristic). Vowel *groups* ≈ syllables
// for Latin/romanized text; for non-Latin scripts (CJK, Devanagari, …) each
// letter is roughly one mora, so letter count is the better estimate.
const LATIN_RE = /[a-z]/i;
export function syllableCount(word) {
  const w = (word || '').toLowerCase();
  if (LATIN_RE.test(w)) {
    let n = (w.match(/[aeiouyàáâäãåèéêëìíîïòóôöõùúûüỳýŷÿ]+/g) || []).length;
    // Drop a silent trailing "e" (e.g. "time" → 1, not 2).
    if (n > 1 && /[^aeiou]e\b/.test(w)) n -= 1;
    return Math.max(1, n);
  }
  const letters = [...w].filter((c) => /\p{L}/u.test(c)).length;
  return Math.max(1, letters);
}

const LINE_RE = /^((?:\[\d{1,2}:\d{1,2}(?:\.\d{1,3})?\])+)(.*)$/;
const TAG_RE = /\[(\d{1,2}):(\d{1,2}(?:\.\d{1,3})?)\]/g;
const WORD_TS_RE = /<(\d{1,2}):(\d{1,2}(?:\.\d{1,3})?)>/g;

function toSeconds(mm, ss) {
  return parseInt(mm, 10) * 60 + parseFloat(ss);
}

function parseRawLines(lrc) {
  const out = [];
  for (const raw of lrc.split(/\r?\n/)) {
    const m = raw.match(LINE_RE);
    if (!m) continue;
    const [, stamps, body] = m;

    let tagMatch;
    TAG_RE.lastIndex = 0;
    const starts = [];
    while ((tagMatch = TAG_RE.exec(stamps)) !== null) {
      starts.push(toSeconds(tagMatch[1], tagMatch[2]));
    }

    const text = body.replace(WORD_TS_RE, '').trim();
    const wordTimes = extractWordTimes(body);
    for (const start of starts) {
      out.push({ start, text, wordTimes: wordTimes.length ? wordTimes : null });
    }
  }
  out.sort((a, b) => a.start - b.start);
  return out;
}

function extractWordTimes(body) {
  const times = [];
  let m;
  WORD_TS_RE.lastIndex = 0;
  while ((m = WORD_TS_RE.exec(body)) !== null) {
    times.push(toSeconds(m[1], m[2]));
  }
  return times;
}

export function parseLRC(lrc, { trailingLineSeconds = 4 } = {}) {
  const rawLines = parseRawLines(lrc).filter((l) => l.text.length > 0);
  if (!rawLines.length) return { lines: [], duration: 0 };

  const lines = rawLines.map((line, i) => {
    const next = rawLines[i + 1];
    const end = next ? next.start : line.start + trailingLineSeconds;
    const tokens = line.text.split(/\s+/).filter(Boolean);

    let words;
    if (line.wordTimes && line.wordTimes.length === tokens.length) {
      words = tokens.map((text, wi) => ({
        text,
        start: line.wordTimes[wi],
        end: wi + 1 < line.wordTimes.length ? line.wordTimes[wi + 1] : end,
      }));
    } else {
      // No per-word timestamps: estimate them. Shared with the plain/estimated and
      // ASR paths so every "words estimated" source uses one corrected model.
      words = wordsAcrossSpan(tokens, line.start, end);
    }
    return { start: line.start, end, words };
  });

  return { lines, duration: lines[lines.length - 1].end };
}

File: tvos/Bar4BarCore/Sources/Bar4BarCore/Formats/LRC.swift
import Foundation

public enum LRC {
  private static let lineRE = try! NSRegularExpression(
    pattern: #"^((?:\[\d{1,2}:\d{1,2}(?:\.\d{1,3})?\])+)(.*)$"#
  )
  private static let tagRE = try! NSRegularExpression(
    pattern: #"\[(\d{1,2}):(\d{1,2}(?:\.\d{1,3})?)\]"#
  )
  private static let wordTsRE = try! NSRegularExpression(
    pattern: #"<(\d{1,2}):(\d{1,2}(?:\.\d{1,3})?)>"#
  )

  private static func toSeconds(mm: String, ss: String) -> Double {
    (Double(mm) ?? 0) * 60 + (Double(ss) ?? 0)
  }

  private struct RawLine {
    var start: Double
    var text: String
    var wordTimes: [Double]?
  }

  public static func parse(_ lrc: String, trailingLineSeconds: Double = 4) -> Timeline {
    let rawLines = parseRawLines(lrc).filter { !$0.text.isEmpty }
    guard !rawLines.isEmpty else { return Timeline(lines: [], duration: 0) }

    var estimated = false
    let lines: [LyricLine] = rawLines.enumerated().map { i, line in
      let next = i + 1 < rawLines.count ? rawLines[i + 1] : nil
      let end = next?.start ?? (line.start + trailingLineSeconds)
      let tokens = Estimate.tokenizeLine(line.text)
      let words: [LyricWord]
      if let times = line.wordTimes, times.count == tokens.count {
        words = tokens.enumerated().map { wi, text in
          LyricWord(
            text: text,
            start: times[wi],
            end: wi + 1 < times.count ? times[wi + 1] : end,
            timingQuality: .reliable
          )
        }
      } else {
        estimated = true
        words = Estimate.wordsAcrossSpan(tokens: tokens, start: line.start, end: end)
      }
      return LyricLine(start: line.start, end: end, words: words)
    }

    return Timeline(lines: lines, duration: lines.last?.end ?? 0, estimated: estimated, source: estimated ? "lrc" : "elrc")
  }

  private static func parseRawLines(_ lrc: String) -> [RawLine] {
    var out: [RawLine] = []
    for raw in lrc.split(whereSeparator: \.isNewline).map(String.init) {
      let ns = raw as NSString
      let full = NSRange(location: 0, length: ns.length)
      guard let m = lineRE.firstMatch(in: raw, range: full),
            m.numberOfRanges >= 3
      else { continue }
      let stamps = ns.substring(with: m.range(at: 1))
      let body = ns.substring(with: m.range(at: 2))

      var starts: [Double] = []
      tagRE.enumerateMatches(in: stamps, range: NSRange(location: 0, length: (stamps as NSString).length)) { match, _, _ in
        guard let match, match.numberOfRanges >= 3 else { return }
        let s = stamps as NSString
        starts.append(toSeconds(mm: s.substring(with: match.range(at: 1)), ss: s.substring(with: match.range(at: 2))))
      }

      let text = wordTsRE.stringByReplacingMatches(
        in: body,
        range: NSRange(location: 0, length: (body as NSString).length),
        withTemplate: ""
      ).trimmingCharacters(in: .whitespacesAndNewlines)

      let wordTimes = extractWordTimes(body)
      for start in starts {
        out.append(RawLine(
          start: start,
          text: text,
          wordTimes: wordTimes.isEmpty ? nil : wordTimes
        ))
      }
    }
    out.sort { $0.start < $1.start }
    return out
  }

  private static func extractWordTimes(_ body: String) -> [Double] {
    var times: [Double] = []
    let ns = body as NSString
    wordTsRE.enumerateMatches(in: body, range: NSRange(location: 0, length: ns.length)) { match, _, _ in
      guard let match, match.numberOfRanges >= 3 else { return }
      times.append(toSeconds(
        mm: ns.substring(with: match.range(at: 1)),
        ss: ns.substring(with: match.range(at: 2))
      ))
    }
    return times
  }
}
